### DISCUSSION TITLE
在一段时间没有请求某些域名的时候再次请求的第一次会返回nxdomain，但刷新一次后正常

### DISCUSSION THREAD

**Discussion by BA7JCM** [AdguardTeam/AdGuardHome#3537] [Q&A]
Have a question or an idea? Please search it [on our forum](https://github.com/AdguardTeam/AdGuardHome/discussions) to make sure it was not yet asked. If you cannot find what you had in mind, please [submit it here](https://github.com/AdguardTeam/AdGuardHome/discussions/new).

### Prerequisites

Please answer the following questions for yourself before submitting an issue. **YOU MAY DELETE THE PREREQUISITES SECTION.**

- [* ] I am running the latest version
- [* ] I checked the documentation and found no answer
- [* ] I checked to make sure that this issue has not already been filed

### Issue Details

* **Version of AdGuard Home server:**
  * 
  *  v0.107.0-a.150+1265be50
* **How did you install AdGuard Home:**
  * 
  * 使用readme.md中的脚本安装的
* **How did you setup DNS configuration:**
  * 
  * system&router&dns over https（在浏览器中）&dns over tls（在手机上）
* **If it's a router or IoT, please write device model:**
  * 
  * router：xiaomi redme ac2100
* **CPU architecture:**
  * 
  * 电脑是amd64，手机是arm64v8（好像是吧），路由器的CPU架构不详
* **Operating system and version:**
  * 
搭建adgh的服务器：Debian
使用的设备：电脑Windows10，手机Android11（miui12.5）
### Expected Behavior11

所有正确的域名能够正常被解析/返回被拦截的提示
### Actual Behavior

在长时间没有请求过的域名（或者是从未请求过的域名）在第一次请求的时候会返回没有解析，但是刷新一次就能正常解析，另外有个别域名访问的时候提示无法解析，在adgh的解析记录中也没有发现有请求这个域名，但是我把这个域名手动在自定义规则中添加白名单之后就能解析了
### Screenshots

<details><summary>Screenshot:</summary>
例如这个：

![image](https://user-images.githubusercontent.com/79561519/131361686-08f0be0e-de84-41b4-8e60-4be7b19f7918.png)

我第一次访问是解析不出来，但是我刷新了一下

![image](https://user-images.githubusercontent.com/79561519/131361820-c0c42c51-57e9-48f6-b457-400fd9d47424.png)

就能解析出来
</details>

### Additional Information

但是我搭建在从腾讯云租的虚拟服务器（系统是centos7.9）上面搭建的在使用doh/dot时没有遇到过这个问题，adgh版本是 
v0.106.3，但因腾讯云对个人使用53端口有限制，故无法使用非加密的dns，服务器上面的版本是从release中下载后自行解压安装的

**Comment by ameshkov** [comment]
> 在adgh的解析记录中也没有发现有请求这个域名

这很奇怪，你确定这个请求到了AGH吗？

请查看AGH的“verbose”日志：https://github.com/AdguardTeam/AdGuardHome/wiki/FAQ#verboselog
能不能在日志文件找到这些请求？

> 上面搭建的在使用doh/dot时没有遇到过这个问题

这个也很奇怪，有可能有某个UDP问题。

**Reply by BA7JCM** [reply]
> > 在adgh的解析记录中也没有发现有请求这个域名
> 
> 这很奇怪，你确定这个请求到了AGH吗？


不确定，但如果没有请求到ADgh，那为啥我在adgh中手动添加为白名单后就能解析了


> > 上面搭建的在使用doh/dot时没有遇到过这个问题
> 
> 这个也很奇怪，有可能有某个UDP问题。


但是我搭建在本地的adgh我也配置了dot/doh，且浏览器连接它也是走的doh，与我连接远端服务器上的adgh的方式完全相同

**Reply by ameshkov** [reply]
我希望日志文件会帮助我们了解这个情况。而且，我们需要知道什么域名有问题。

**Reply by BA7JCM** [reply]
> 我希望日志文件会帮助我们了解这个情况。而且，我们需要知道什么域名有问题。

已启用详细记录并设置日志文件路径，目前正在观察看问题是否再次出现，另外在使用了adgh之后多了很多原本可以解析的域名都无法解析，有的只需刷新一下就能解析，有的则需要进后台手动添加白名单才能解析

**Reply by BA7JCM** [reply]
> 我希望日志文件会帮助我们了解这个情况。而且，我们需要知道什么域名有问题。

[QQ录屏20210902220927.zip](https://github.com/AdguardTeam/AdGuardHome/files/7099530/QQ.20210902220927.zip)
[adgh.log](https://github.com/AdguardTeam/AdGuardHome/files/7099573/adgh.log)
刚刚成功复现问题“域名在访问的时候需要刷新一次才能正常解析”，已附上录屏文件和adgh的日志

**Reply by BA7JCM** [reply]
> 我希望日志文件会帮助我们了解这个情况。而且，我们需要知道什么域名有问题。

而且出现这个问题的时候大部分都是点击超链接跳转的时候和一个页面里面调用其它域名的东西的时候，会马上显示无法解析，没有等待时间，但是我刷新页面以后又能解析出来

**Reply by BA7JCM** [reply]
> 我希望日志文件会帮助我们了解这个情况。而且，我们需要知道什么域名有问题。

我刚刚翻了一下adgh的解析记录，发现出问题的域名都是第一次请求的时候返回“noerror”代码但没有返回别的东西，如图所示：
![image](https://user-images.githubusercontent.com/79561519/132035723-33506ae1-8c32-4e20-b2f6-6515e9292f56.png)

要第二次请求才会返回解析的IP，如图所示：
![image](https://user-images.githubusercontent.com/79561519/132035731-619625d3-79c9-4fde-8566-eb1bc318a0d7.png)

我发现这个问题存在于几乎大部分域名上面，只有少数的一些域名是可以直接“一次过”解析出来的

**Reply by ameshkov** [reply]
在你上传的日志里，上游服务器回答 SERVFAIL：
```
2021/09/02 22:08:40.844663 431330#2522 [debug] github.com/AdguardTeam/dnsproxy/proxy.(*Proxy).logDNSMessage(): OUT: ;; opcode: QUERY, status: SERVFAIL, id: 0
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; QUESTION SECTION:
;www.debian.org.	IN	 AAAA

;; ADDITIONAL SECTION:

;; OPT PSEUDOSECTION:
; EDNS: version 0; flags: do; udp: 4096
```

```
2021/09/02 22:08:40.846360 431330#2562 [debug] github.com/AdguardTeam/dnsproxy/proxy.(*Proxy).logDNSMessage(): OUT: ;; opcode: QUERY, status: SERVFAIL, id: 0
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1

;; QUESTION SECTION:
;gravatar.wp-china-yes.net.	IN	 AAAA

;; ADDITIONAL SECTION:

;; OPT PSEUDOSECTION:
; EDNS: version 0; flags: do; udp: 4096
```

**Reply by BA7JCM** [reply]
> 在你上传的日志里，上游服务器回答 SERVFAIL：
> ```
> 2021/09/02 22:08:40.844663 431330#2522 [debug] github.com/AdguardTeam/dnsproxy/proxy.(*Proxy).logDNSMessage(): OUT: ;; opcode: QUERY, status: SERVFAIL, id: 0
> ;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
> 
> ;; QUESTION SECTION:
> ;www.debian.org.	IN	 AAAA
> 
> ;; ADDITIONAL SECTION:
> 
> ;; OPT PSEUDOSECTION:
> ; EDNS: version 0; flags: do; udp: 4096
> ```
> 
> ```
> 2021/09/02 22:08:40.846360 431330#2562 [debug] github.com/AdguardTeam/dnsproxy/proxy.(*Proxy).logDNSMessage(): OUT: ;; opcode: QUERY, status: SERVFAIL, id: 0
> ;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
> 
> ;; QUESTION SECTION:
> ;gravatar.wp-china-yes.net.	IN	 AAAA
> 
> ;; ADDITIONAL SECTION:
> 
> ;; OPT PSEUDOSECTION:
> ; EDNS: version 0; flags: do; udp: 4096
> ```

但是我设置了不止一个上游DNS服务器，而且也选择了遍历所有DNS服务器，为啥还是会出现这种情况，而且为啥我刷新一下之后就能解析了，我该如何解决这个问题

**Reply by ameshkov** [reply]
> 但是我设置了不止一个上游DNS服务器

AGH 不检查回应，简单的交给客户。如果你用 “最快的 IP 地址”，AGH就检查所有的回应，试试 IP。

> 而且为啥我刷新一下之后就能解析了

有可能下个请求它用别的服务器。

**Reply by BA7JCM** [reply]
> > 但是我设置了不止一个上游DNS服务器
> 
> AGH 不检查回应，简单的交给客户。如果你用 “最快的 IP 地址”，AGH就检查所有的回应，试试 IP。
> 
> > 而且为啥我刷新一下之后就能解析了
> 
> 有可能下个请求它用别的服务器。

我尝试了“使用最快IP”，处理请求的耗时由原来的数百毫秒变为了一万多毫秒。。。。。。另外我在adgh的管理页面发现有时候的请求只会返回一个“noerror”响应代码而没有返回响应内容，在刷新后才返回响应内容
